### PR TITLE
add fastpaths for Zeros and Ones

### PR DESCRIPTION
**Pull request by CarloLucibello**
add fast paths for broadcasted additions and multiplications involving Zeros and Ones

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -885,7 +885,7 @@ end
     @test A*Fill(2,4,1) == A*Matrix(Fill(2,4,1))
     @test a*Fill(2,1,3) == a*Matrix(Fill(2,1,3))
     @test A*Ones(4) ==   A*Vector(Ones(4))
-    @test A*Ones(4,1) == A*Matrix(Ones(4,1))
+    @test A*Ones(4,1) ≈ A*Matrix(Ones(4,1))
```
Comment (CarloLucibello) [inline test/runtests.jl:888]: add to change this since it was failing on my machine

File: src/fillbroadcast.jl [hunk 2]
```diff
@@ -93,3 +93,64 @@ broadcasted(::DefaultArrayStyle{N}, op, r::AbstractFill{T,N}, x::Number) where {
...
+        size(a) == bs && return a
+        c = similar(a, bs)
+        c .= a
+        return c
+    end
+
+    @eval function broadcasted(::DefaultArrayStyle, ::typeof($op), 
+                                a::AbstractFill, b::Zeros)
+        bs = broadcast_shape(size(a), size(b))
+        Fill(getindex_value(a),  bs)
```
Comment (dlfivefifty) [inline src/fillbroadcast.jl:111]: Use `fillsimilar(a, bs)` so that `Ones` stays a `Ones`.

File: src/fillbroadcast.jl [hunk 3]
```diff
@@ -93,3 +93,64 @@ broadcasted(::DefaultArrayStyle{N}, op, r::AbstractFill{T,N}, x::Number) where {
...
+    bs = broadcast_shape(size(a), size(b))
+    size(b) == bs && return b
+    c = similar(b, bs)
+    c .= b
+    return c
+end
+
+function broadcasted(::DefaultArrayStyle, ::typeof(+), a::Zeros, b::AbstractFill)
+    bs = broadcast_shape(size(a), size(b))
+    Fill(getindex_value(b),  bs)
```
Comment (dlfivefifty) [inline src/fillbroadcast.jl:125]: Use `fillsimilar`

File: src/fillbroadcast.jl [hunk 4]
```diff
@@ -93,3 +93,64 @@ broadcasted(::DefaultArrayStyle{N}, op, r::AbstractFill{T,N}, x::Number) where {
...
+        size(a) == bs && return a
+        c = similar(a, bs)
+        c .= a
+        return c
+    end
+
+    @eval function broadcasted(::DefaultArrayStyle, ::typeof($op), 
+                                a::AbstractFill, b::Ones)
+        bs = broadcast_shape(size(a), size(b))
+        Fill(getindex_value(a),  bs)
```
Comment (dlfivefifty) [inline src/fillbroadcast.jl:141]: Use `fillsimilar`

File: test/runtests.jl [hunk 5]
```diff
@@ -885,7 +885,7 @@ end
     @test A*Fill(2,4,1) == A*Matrix(Fill(2,4,1))
```
Comment (dlfivefifty) [inline test/runtests.jl:885]: Add tests that failed before but now pass.

File: test/runtests.jl [hunk 6]
```diff
@@ -564,6 +564,37 @@ end
...
+        @test Zeros() .+ x !== x
+        @test Zeros(2) .+ x !== x
+        @test x .* Ones() !== x
+        @test x .* Ones(2)  !== x
+        @test x ./ Ones() !== x
+        @test x ./ Ones(2)  !== x
+        @test x .+ Zeros()  !== x
+        @test x .+ Zeros(2)  !== x
+        @test x .- Zeros()  !== x
+        @test x .- Zeros(2)  !== x
```
Comment (dlfivefifty) [inline test/runtests.jl:596]: Add tests for `AbstractFill`, e.g., `Ones(2) .* Zeros(2)`

Comment (CarloLucibello) [reply test/runtests.jl]: done, along with some fixes

**Comment by CarloLucibello** [comment]
this PR was quite some combinatorial gymnastic, but I think it is ready now

**Comment by dlfivefifty** [comment]
We also need tests for the special case broadcasting involving row or column vectors, e.g.,:
```julia
julia> Ones(5) .+ Ones(1,5)
5×5 Fill{Float64,2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}}:
 2.0  2.0  2.0  2.0  2.0
 2.0  2.0  2.0  2.0  2.0
 2.0  2.0  2.0  2.0  2.0
 2.0  2.0  2.0  2.0  2.0
 2.0  2.0  2.0  2.0  2.0
```

**Comment by CarloLucibello** [comment]
I tried to come up with fewer and more generic methods, specializing on `AbstractFill` and using `fillsimilar` for the return type, but this led to many method ambiguities that I could not overcome

**Comment by CarloLucibello** [comment]
Ah, pity, this got forgotten, I'll rebase when I'll find some time

**Comment by CarloLucibello** [comment]
rebase was too messy, I'll file a new PR

**Comment by dlfivefifty** [comment]
Should we close this one then?

**Pull request closed without merging**